### DISCUSSION TITLE
What is a reason to distinguish preprocessor definitions?

### DISCUSSION THREAD

**Discussion by lolgear** [libgit2/libgit2#6249] [Q&A]
Recent release ( 1.3.0 ) still separates macro definitions into two categories: features and others.
First category gathers definitions into features.h header.
Second category adds definitions as preprocessor directives to target in cmake build step.
Could we eliminate second category and put every definition into features.h header?

I guess that I'm missing something.

**Comment by lolgear** [comment]
@ethomson Could you comment?

**Reply by ethomson** [reply]
Can you give me an example of something in the second category? I guess I'm not quite sure of the distinction that you're trying to draw yet.

**Reply by lolgear** [reply]
Sure.

First category can be described by features.h file.

```
/* #undef GIT_SSH */
/// and many more
```

Second category can be described by configured xcodeproj file ( if you use "Xcode" generator ).

`git2internal` target contains

```
//:configuration = Debug
GCC_PREPROCESSOR_DEFINITIONS = 'CMAKE_INTDIR="$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)"' HAVE_QSORT_R_BSD '_FILE_OFFSET_BITS=64' 'SHA1DC_NO_STANDARD_INCLUDES=1' 'SHA1DC_CUSTOM_INCLUDE_SHA1_C="common.h"' 'SHA1DC_CUSTOM_INCLUDE_UBC_CHECK_C="common.h"' $(inherited)
```

Here we can see

```
HAVE_QSORT_R_BSD
_FILE_OFFSET_BITS=64
SHA1DC_NO_STANDARD_INCLUDES=1
SHA1DC_CUSTOM_INCLUDE_SHA1_C="common.h"
SHA1DC_CUSTOM_INCLUDE_UBC_CHECK_C="common.h"
```

But all of them can be inserted into `features.h`, no?

**Reply by lolgear** [reply]
@ethomson Could you comment?

**Reply by ethomson** [reply]
features.h is a thing that we expose to callers. it's about the feature set that is compiled into libgit2. some people may want to care about whether the library is using openssl or securetransport, or if it has ntlm support. if anything, there are some things that i would _remove_ from `features.h` - like qsort mechanics. but nobody should know about `SHA1DC_CUSTOM_INCLUDE_SHA1_C="common.h"`, that's a compile-time detail that should remain private to the library.

**Reply by lolgear** [reply]
It would be nice to have an option "output definitions to a file" or "output private definitions to a features.h".
For example, Xcode supports two project files "xcodeproj" and "SPM manifest file".
But cmake supports only "xcodeproj" generator which is very inconvenient for library-only targets.
So, if you would like to support SPM, you have to check whether private definitions change or not.
